File: LeetCode/21. Merge Two Sorted Lists.py
# https://leetcode.com/problems/merge-two-sorted-lists/

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        h1 = l1
        h2 = l2
        
        if h1==None and h2==None:
            return None
        result = []
        result_n = ListNode() 
        while(h1 != None or h2 != None):
            if h1 != None:
                if h2 !=None and h1.val<=h2.val:
                    result.append(h1.val)
                    h1 = h1.next
                elif h2==None:
                    result.append(h1.val)
                    h1 = h1.next
            if h2 != None:
                if h1 !=None and h2.val<=h1.val:
                    result.append(h2.val)
                    h2 = h2.next
                elif h1==None:
                    result.append(h2.val)
                    h2 = h2.next
        rn = result_n
        for i in range(0, len(result)):
            result_n.val = result[i]
            if i < len(result)-1:
                result_n.next = ListNode(result[i+1])
            result_n = result_n.next
        return rn 

File: LeetCode/21. Merge Two Sorted Lists.go
// https://leetcode.com/problems/merge-two-sorted-lists/

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
    var dummy ListNode
    // tail will like pointer. It will move from beginning until last node.
    // we use address of dummy to start the tail so dummy can follow the path without moving
    tail := &dummy
    
    for list1 != nil && list2 != nil {
        // next value is lowest head in two linked list
        // and then move head to next node
        if list1.Val < list2.Val {
            tail.Next = list1
            list1 = list1.Next
        } else {
            tail.Next = list2
            list2 = list2.Next
        }
        // move tail to next node
        tail = (*tail).Next
    }
    
    // if len is not same, get latest data to link list
    if list1 != nil {
        tail.Next = list1
    } else if list2 != nil {
        tail.Next = list2
    }
    return dummy.Next
}
